yew/src/html/classes.rs

```diff
@@ -0,0 +1,641 @@
+use indexmap::IndexSet;
+use std::{
+    borrow::{Borrow, Cow},
+    iter::FromIterator,
+};
+
+/// A set of classes.
+#[derive(Debug, Clone, Default)]
+pub struct Classes {
+    set: IndexSet<Cow<'static, str>>,
+}
+
+impl Classes {
+    /// Creates an empty set of classes. (Does not allocate.)
+    pub fn new() -> Self {
+        Self {
+            set: IndexSet::new(),
+        }
+    }
+
+    /// Creates an empty set of classes with capacity for n elements. (Does not allocate if n is
+    /// zero.)
+    pub fn with_capacity(n: usize) -> Self {
+        Self {
+            set: IndexSet::with_capacity(n),
+        }
+    }
+
+    /// Adds a class to a set.
+    ///
+    /// If the provided class has already been added, this method will ignore it.
+    pub fn push<T: Into<Self>>(&mut self, class: T) {
+        let classes_to_add: Self = class.into();
+        self.set.extend(classes_to_add.set);
+    }
+
+    /// Adds a class to a set.
+    ///
+    /// If the provided class has already been added, this method will ignore it.
+    ///
+    /// This method won't check if there are multiple classes in the input string.
+    pub fn unchecked_push<T: Into<Cow<'static, str>>>(&mut self, class: T) {
```

Comment:
This method should be unsafe because it requires the user to manually uphold an invariant.
```suggestion
    pub unsafe fn unchecked_push<T: Into<Cow<'static, str>>>(&mut self, class: T) {
```

Reply:
I'm not sure I understand :thinking:

Reply:
Why this should be `unsafe`? It's for the same reason that something like `String::from_utf8_unchecked` is unsafe. It's not because the function itself can cause memory corruption or a crash, but because the function can be used to accidentally break an invariant. In the case of `String` this is that all strings need to be valid UTF8 and `from_utf8_unchecked` doesn't have a check in place to ensure that this is true.
In the case of `Classes` the invariant is that it's a list of separate classes. We enforce this by always splitting strings on whitespace, right? But this function doesn't do that and so it can be used to break this invariant. To make this clear to the user (like a little nudge to read the documentation) we mark it as unsafe.

Reply:
Makes sense. Thanks!! :heart: